### ISSUE TITLE
Support "maximum length" validation

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#2242]
This only addresses the "maximum length" and possibly "minimum length" validations, but we would like to do it in a way that allow working towards more validations and combinators (#1112).

We are leaning towards using attributes and later children elements to represent declaratively a validation expression.

Questions:

- should `xxf:max-length` support a non-static value?
- should we convert `xxf:max-length` into an XPath expression at compilation time?
  - could then be transformed into `string-length(.) le ($value)` or probably better `xxf:max-length(., ($value))`

Possibility:

- answer yes to both questions above
- convert the attribute to an XPath expression
- combine the expression with an existing `constraint` attribute if present
- `xxf:max-length()` stores in the context, during execution, the value of its second operand (the max length)
- the evaluation of the resulting constraint will be
  - a boolean (whether the constraint was satisfied)
  - if `xxf:max-length()` ran, the "effective max length" requested
- the "effective max length" value, if present, but be stored into the instance data node, otherwise removed

*NOTE/QUESTION: Right now, constraints combine with `and`, and each constraint can have its own alert. Do we need short-circuiting for `and` in general, or not? The validation tree could possibly encompass general XPath expressions, and each top-level `and` operand could have an associated error message.*

**Comment by ebruchez** [comment]
We are now leaning toward using a subset of XPath. Arguments:

- just specifying `xxf:max-length` as attribute/element is easy enough
- BUT in the future we want richer validation with conjunctions/disjunctions
- this asks the question of the syntax of the expression tree represented probably in XML
  - we might need to use the `xf:constraint` child element or another element name
- the XForms engine
  - must parse and store that representation
  - must interpret that validation tree

In short:

1. We don't like the idea of specifying yet another XML sub-language. We think that using a subset of XPath within `constraint` and `xf:constraint` might be better.
2. Using a subset of XPath still requires changes, but less at the XForms engine level.

**Comment by ebruchez** [comment]
NOTE: This issue also covers the "minimum length" validation.
